### PR TITLE
Set location response header to the final destination of all participating automatic transitions

### PR DESCRIPTION
**Pull request by dlgroves**
This change updates IRIS's location header to the destination of a fully resolved chain of automatic transitions and appends any parameters that were resolved during this process.

### CODE REVIEW

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java [hunk 1]
```diff
@@ -650,25 +650,27 @@ private Response buildResponse(HttpHeaders headers, MultivaluedMap<String, Strin
...
             List<Transition> autoTransitions = getTransitions(ctx, currentState, Transition.AUTO);
-
             if (!autoTransitions.isEmpty()) {
-                Transition autoTransition = autoTransitions.get(0);
-                if (autoTransitions.size() > 1)
-                    logger.warn("Resource state [" + currentState.getName()
+                List<Transition> innerTransitions = autoTransitions;
+                Transition autoTransition;
+                do {
+                    autoTransition = innerTransitions.get(0);
```
Comment (mtemenos) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: Given innerTransitions has to have at least one element why not simplify the code by using a forach loop?

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java [hunk 2]
```diff
@@ -681,31 +683,26 @@ private Response buildResponse(HttpHeaders headers, MultivaluedMap<String, Strin
...
-                    if (currentState != tmpTransition.getTarget()) {
-                        autoTransition = tmpTransition;
-                    }
-                }
-
-                if (autoTransition != null) {
-                    if (autoTransitions.size() > 1)
+                List<Transition> innerTransitions = autoTransitions;
+                Transition autoTransition;
+                do {
```
Comment (mtemenos) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: Same point as above for the code above

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java [hunk 3]
```diff
@@ -627,13 +627,13 @@ private Response buildResponse(HttpHeaders headers, MultivaluedMap<String, Strin
                 if (autoTransitions.size() > 1)
                     logger.warn("Resource state [" + currentState.getName()
                             + "] has multiple auto-transitions. Using [" + autoTransition.getId() + "].");
-                Response autoResponse = getResource(headers, autoTransition, ctx);
```
Comment (mtemenos) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: There is way to much going on in build response. Please can you extract methods for the areas you are changing.

Comment (dlgroves) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: I agree; logic for automatic transition handling has been refactored into a separate method.

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java (hunk 1 above)
Comment (dlgroves) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: I disagree; this code is traversing a graph of automatic transitions, not iterating over a list.

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java (hunk 2 above)
Comment (dlgroves) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: See above and below.

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java (hunk 3 above)
Comment (mtemenos) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HTTPHypermediaRIM.java]: Great - Thank you!

File: interaction-core/src/main/java/com/temenos/interaction/core/rim/HeaderHelper.java [hunk 4]
```diff
@@ -22,13 +22,23 @@
...
+import java.util.Map;
 import java.util.Set;
 
 import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MultivaluedMap;
 import javax.ws.rs.core.Response.ResponseBuilder;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (mtemenos) [inline interaction-core/src/main/java/com/temenos/interaction/core/rim/HeaderHelper.java]: A brief class summary would be good e.g. "This class provides helper functions for manipulating HTTP headers" Even better might be to rename the class HttpHeaderHelper?

File: interaction-core/src/main/java/com/temenos/interaction/core/hypermedia/LinkGeneratorImpl.java [hunk 5]
```diff
@@ -54,7 +54,7 @@
 public class LinkGeneratorImpl implements LinkGenerator {
 
     protected static final String PARAM_REPLACEMENT_REGEX = "\\((\\d+)\\)";
-    private final Logger logger = LoggerFactory.getLogger(LinkGeneratorImpl.class);
+    private static final Logger logger = LoggerFactory.getLogger(LinkGeneratorImpl.class);
     private ResourceStateMachine resourceStateMachine;
```
Comment (mtemenos) [inline interaction-core/src/main/java/com/temenos/interaction/core/hypermedia/LinkGeneratorImpl.java]: LinkGeneratorImpl doesn't have a unit test - We should note to add one.

**Pull request merged**